### PR TITLE
make cachesync timeout configurable

### PR DESCRIPTION
**Pull request by listx**
Tide, Horologium, and Deck all call WaitForCacheSync [1]. However only Deck allows the timeout to be configurable through its own `timeoutListingProwJobs` (--timeout-listing-prowjobs) flag.

Here we promote this flag to be sharable across other components. Tide and Horologium both make use of this new flag. Also, the default value has been raised to 60 seconds (up from 30 seconds). This new 60 second timeout affects both Deck and Tide, which used to have a 30 second timeout.

Also, we replace the `cacheSyncCtx` and `cacheSyncCancel` variable names in Horologium with `mgrSyncCtx` and `mgrSyncCtxCancel`, respectively, for consistency.

[1]: https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.6/pkg/cache#Informers.WaitForCacheSync

/cc @cjwagner

### CODE REVIEW

File: prow/flagutil/config/config.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ type ConfigOptions struct {
 	// Inrepoconfig related flags
 	InRepoConfigCacheSize    int
 	InRepoConfigCacheDirBase string
+	// Controller runtime manager
+	TimeoutListingProwJobs int
```
Comment (cjwagner) [inline prow/flagutil/config/config.go:47]: This is not related to configuration at all. Please add a new flag bundle under flagutil rather than reusing `ConfigOptions` even if it is just the one flag we're adding.

File: prow/flagutil/config/config.go [hunk 2]
```diff
@@ -59,6 +61,7 @@ func (o *ConfigOptions) AddFlags(fs *flag.FlagSet) {
 	fs.StringVar(&o.SupplementalProwConfigsFileNameSuffix, "supplemental-prow-configs-filename-suffix", "_prowconfig.yaml", "Suffix for additional prow configs. Only files with this name will be considered")
 	fs.IntVar(&o.InRepoConfigCacheSize, "in-repo-config-cache-size", 1000, "Cache size for ProwYAMLs read from in-repo configs. Each host receives its own cache.")
 	fs.StringVar(&o.InRepoConfigCacheDirBase, "cache-dir-base", "", "Directory where the repo cache should be mounted.")
+	fs.IntVar(&o.TimeoutListingProwJobs, "timeout-listing-prowjobs", 60, "Timeout for listing prowjobs in seconds.")
```
Comment (cjwagner) [inline prow/flagutil/config/config.go:64]: I don't think we should be changing the default value to 60. 
- I don't think it will break anyone, but it is possible that it could affect deployment strategy configuration. 
- More importantly, 30 seconds is a more reasonable default. We're only raising this to 60 seconds in our use case as a temporary performance mitigation.

Comment (cjwagner) [inline prow/flagutil/config/config.go:64]: Please use a duration type flag rather than an int type with implied units.

Comment (cjwagner) [reply prow/flagutil/config/config.go:64]: It looks like we actually had a different default for horologium. If we want to keep this we could add a default flag value field to the new flag bundle (mentioned in my other comment) like we do here with [`NOInClusterConfigDefault`](https://github.com/kubernetes/test-infra/blob/1dcc1a65ba4b4f47b19fcca8121f38256164bbe5/prow/flagutil/kubernetes_cluster_clients.go#L65).

Comment (listx) [reply prow/flagutil/config/config.go:64]: The example you linked has 2 fields:
```
	noInClusterConfig        bool
	NOInClusterConfigDefault bool
```

so are you suggesting that we can have the same thing here for TimeoutListingProwJobs? (I.e., add a second TimeoutListingProwJobsDefault field?)

Comment (cjwagner) [reply prow/flagutil/config/config.go:64]: Yes, the first is the actual flag variable and the second (with the "Default" suffix) is the default value to use for the flag. We can then override the default flag value of 30 in the main function of horologium by setting `.TimeoutListingProwJobsDefault = 60` before we add the flags.

File: prow/cmd/deck/main.go [hunk 3]
```diff
@@ -190,11 +190,12 @@ func gatherOptions(fs *flag.FlagSet, args ...string) options {
 	fs.BoolVar(&o.gcsCookieAuth, "gcs-cookie-auth", false, "Use storage.cloud.google.com instead of signed URLs")
 	fs.BoolVar(&o.rerunCreatesJob, "rerun-creates-job", false, "Change the re-run option in Deck to actually create the job. **WARNING:** Only use this with non-public deck instances, otherwise strangers can DOS your Prow instance")
 	fs.BoolVar(&o.allowInsecure, "allow-insecure", false, "Allows insecure requests for CSRF and GitHub oauth.")
-	fs.IntVar(&o.timeoutListingProwJobs, "timeout-listing-prowjobs", 30, "Timeout for listing prowjobs in seconds.")
 	fs.BoolVar(&o.dryRun, "dry-run", false, "Whether or not to make mutating API calls to GitHub.")
 	fs.Var(&o.tenantIDs, "tenant-id", "The tenantID(s) used by the ProwJobs that should be displayed by this instance of Deck. This flag can be repeated.")
 	o.config.AddFlags(fs)
 	o.instrumentation.AddFlags(fs)
+	o.controllerManager.TimeoutListingProwJobsDefault = 30
```
Comment (cjwagner) [inline prow/cmd/deck/main.go:197]: This actually sets the duration to 30 nanoseconds I think (or something else very small). This is later adjusted for on line 377 by multiplying by `time.Second`, but it would better to set the value correctly here by setting it to `30*time.Second`. This is how we get benefit from using a duration variable rather than fixing the units to seconds and using an int var. Please fix this in all locations.

aside: I'm surprised there isn't a compiler or linter warning for setting a duration variable to a literal without including a `time.*` constant multiplier.

File: prow/flagutil/controller_manager.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+type ControllerManagerOptions struct {
+	TimeoutListingProwJobs        time.Duration
+	TimeoutListingProwJobsDefault time.Duration
+}
+
+func (o *ControllerManagerOptions) AddFlags(fs *flag.FlagSet) {
+	fs.DurationVar(&o.TimeoutListingProwJobs, "timeout-listing-prowjobs", o.TimeoutListingProwJobsDefault, "Timeout for listing prowjobs in seconds.")
+}
+
+func (o *ControllerManagerOptions) Validate(_ bool) error {
```
Comment (cjwagner) [inline prow/flagutil/controller_manager.go:34]: It looks like you still need to include this in the `Validate` functions for the various component option structs. This isn't actually being used yet. e.g. https://github.com/kubernetes/test-infra/pull/29215/files#diff-d4aa7c79b95cf0dba66f4f832397b3e303120cfb0559753aa812574cd4ec73f1R86

**Review by cjwagner — CHANGES_REQUESTED** [review]
This looks pretty good to me, just a couple small issues.

File: prow/flagutil/controller_manager.go (hunk 4 above)
Comment (listx) [reply prow/flagutil/controller_manager.go:34]: I've made a separate PR for the flag bits because I saw an opportunity to clean things up a bit for consistency. Please review https://github.com/kubernetes/test-infra/pull/29234 first. Then I'll update this PR to make use of the new mechanism there to make sure we call this Validate() function here for each of the components that use it.

File: prow/flagutil/controller_manager.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	"time"
+)
+
+type ControllerManagerOptions struct {
+	TimeoutListingProwJobs        time.Duration
+	TimeoutListingProwJobsDefault time.Duration
+}
+
+func (o *ControllerManagerOptions) AddFlags(fs *flag.FlagSet) {
+	fs.DurationVar(&o.TimeoutListingProwJobs, "timeout-listing-prowjobs", o.TimeoutListingProwJobsDefault, "Timeout for listing prowjobs in seconds.")
```
Comment (cjwagner) [inline prow/flagutil/controller_manager.go:31]: This is not "in seconds".

**Review by cjwagner — APPROVED** [review]
One last minor thing I noticed.
/hold

File: prow/flagutil/controller_manager.go (hunk 5 above)
Comment (listx) [reply prow/flagutil/controller_manager.go:31]: Will fix

Comment (listx) [reply prow/flagutil/controller_manager.go:31]: Thinking about this a little more, it feels very odd that we'll be telling our users to specify this as a pure time.Duration (nanosecond precision) value at runtime. IOW it feels wrong to expect users to say `--timeout-listing-prowjobs=10000000000` to mean 10 seconds. I think I should use a different type (probably just uint) and deal with the type casts.

Cf. https://pkg.go.dev/time#Duration

Comment (cjwagner) [reply prow/flagutil/controller_manager.go:31]: Fortunately that is not how `DurationVar` works and we are not expecting our users to specify a nonsecond precision value, but rather a human readable duration: https://pkg.go.dev/flag#FlagSet.DurationVar

Comment (listx) [reply prow/flagutil/controller_manager.go:31]: TIL. I had a sneaky feeling that there just had to be a better way --- and I'm glad my uint idea is completely unnecessary!

Comment (listx) [reply prow/flagutil/controller_manager.go]: Done

**Review by cjwagner — APPROVED** [review]
